tls/s2n_ktls.c

```diff
@@ -15,11 +15,160 @@
+            RESULT_ENSURE(s2n_stuffer_data_available(&conn->in) == 0, S2N_ERR_RECORD_STUFFER_NEEDS_DRAINING);
+            break;
+        default:
+            RESULT_BAIL(S2N_ERR_SAFETY);
+            break;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_retrieve_file_descriptor(struct s2n_connection *conn, s2n_ktls_mode ktls_mode, int *fd)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(fd);
+
+    if (ktls_mode == S2N_KTLS_MODE_RECV) {
+        RESULT_GUARD_POSIX(s2n_connection_get_read_fd(conn, fd));
+    } else if (ktls_mode == S2N_KTLS_MODE_SEND) {
+        RESULT_GUARD_POSIX(s2n_connection_get_write_fd(conn, fd));
+    }
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_ktls_configure_socket(struct s2n_connection *conn, s2n_ktls_mode ktls_mode)
+{
+    RESULT_ENSURE_REF(conn);
+
+    /* If already enabled then return success */
+    if (ktls_mode == S2N_KTLS_MODE_SEND && conn->ktls_send_enabled) {
+        RESULT_BAIL(S2N_ERR_KTLS_ALREADY_ENABLED);
+    }
+    if (ktls_mode == S2N_KTLS_MODE_RECV && conn->ktls_recv_enabled) {
+        RESULT_BAIL(S2N_ERR_KTLS_ALREADY_ENABLED);
+    }
+
+    int fd = 0;
+    RESULT_GUARD(s2n_ktls_retrieve_file_descriptor(conn, ktls_mode, &fd));
+
+    /* Calls to setsockopt require a real socket, which is not used in unit tests. */
+    if (s2n_in_unit_test()) {
+        return S2N_RESULT_OK;
+    }
+
+    /* Enable 'tls' ULP for the socket. https://lwn.net/Articles/730207 */
+    int ret = setsockopt(fd, S2N_SOL_TCP, S2N_TCP_ULP, S2N_TLS_ULP_NAME, S2N_TLS_ULP_NAME_SIZE);
+
+    if (ret != 0) {
+        /* https://man7.org/linux/man-pages/man3/errno.3.html
+         * EEXIST indicates that TCP_ULP has already been enabled on the socket.
+         * This is a noop and therefore safe to ignore.
+         */
+        RESULT_ENSURE(errno == EEXIST, S2N_ERR_KTLS_ULP);
+    }
+
+    return S2N_RESULT_OK;
+}
+
+/*
+ * Since kTLS is an optimization, it is possible to continue operation
+ * by using userspace TLS if kTLS is not supported.
+ *
+ * kTLS configuration errors are recoverable since calls to setsockopt are
+ * non-destructive and its possible to fallback to userspace.
+ */
+int s2n_connection_ktls_enable_send(struct s2n_connection *conn)
+{
+    POSIX_ENSURE_REF(conn);
+
+    if (!s2n_ktls_is_supported_on_platform()) {
+        POSIX_BAIL(S2N_ERR_KTLS_UNSUPPORTED_PLATFORM);
+    }
+
+    POSIX_GUARD_RESULT(s2n_ktls_validate(conn, S2N_KTLS_MODE_SEND));
+    s2n_result res = s2n_ktls_configure_socket(conn, S2N_KTLS_MODE_SEND);
+    if (s2n_errno == S2N_ERR_KTLS_ALREADY_ENABLED) {
+        return S2N_SUCCESS;
+    } else {
+        POSIX_GUARD_RESULT(res);
+    }
+
+    conn->ktls_send_enabled = true;
+
+    return S2N_SUCCESS;
+}
+
+int s2n_connection_ktls_enable_recv(struct s2n_connection *conn)
+{
+    POSIX_ENSURE_REF(conn);
+
+    if (!s2n_ktls_is_supported_on_platform()) {
+        POSIX_BAIL(S2N_ERR_KTLS_UNSUPPORTED_PLATFORM);
+    }
```

Comment:
```suggestion
    POSIX_ENSURE(s2n_ktls_is_supported_on_platform(), S2N_ERR_KTLS_UNSUPPORTED_PLATFORM);
```
Fewer bails, more ensures :)

But why do this separately from s2n_ktls_validate?

Reply:
yea it can prob live in validate. previously I had 2 separate validation fn so it didnt make sense before